packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -2228,15 +2252,32 @@ class LeaderLayer extends ContainerLayer {
   @override
   void attach(Object owner) {
     super.attach(owner);
-    assert(link._leader == null);
+    assert(link._leader != this);
+    assert((){
+      if (link._leader != null) {
+        link._debugPreviousLeaders ??= <LeaderLayer>{};
+        link._debugPreviousLeaders!.add(link._leader!);
+        link._debugScheduleLeadersCleanUpCheck();
+      }
+      return true;
+    }());
```

Comment:
I would move all of this into a `_registerLeader` method on LeaderLink to keep this a little cleaner here. This is all impl. detail of leaderLink.